File: Java/src/_1137_N_th_Tribonacci_Number/Solution.java
package _1137_N_th_Tribonacci_Number;

/**
 * Date: 2021/8/8 11:19
 * Content:
 * 执行用时：0 ms, 在所有 Java 提交中击败了100.00%的用户
 * 内存消耗：35.1 MB, 在所有 Java 提交中击败了70.79%的用户
 */
public class Solution {

    public int tribonacci(int n) {
        if (n == 0) return 0;
        if (n <= 2) return 1;
        int a = 0, b = 1, c = 1, d = 0;
        for (int i = 0; i < n - 2; i++) {
            d = a + b + c;
            a = b;
            b = c;
            c = d;
        }
        return d;
    }

    public static void main(String[] args) {

        System.out.println((new Solution()).tribonacci(4));  // 4
        System.out.println((new Solution()).tribonacci(25));  // 1389537
    }
}

File: Python/_1137_N_th_Tribonacci_Number/Solution.py
# coding=utf-8
# Date: 2021/8/8 11:26


# 执行用时：32 ms, 在所有 Python3 提交中击败了72.08%的用户
# 内存消耗：14.9 MB, 在所有 Python3 提交中击败了40.24%的用户
class Solution:
    def tribonacci(self, n: int) -> int:
        if n == 0:
            return 0
        if n <= 2:
            return 1
        a, b, c, d = 0, 1, 1, 2
        for i in range(n - 2):
            d = a + b + c
            a, b, c = b, c, d
        return d


if __name__ == "__main__":
    print(Solution().tribonacci(4))  # 4
    print(Solution().tribonacci(25))  # 1389537
